api/src/main/java/io/strimzi/api/kafka/model/kafka/cruisecontrol/HashLoginServiceApiUsers.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model.kafka.cruisecontrol;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import io.strimzi.api.kafka.model.common.Constants;
+import io.sundr.builder.annotations.Buildable;
+import lombok.EqualsAndHashCode;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Cruise Control's API users config
+ */
+@Buildable(
+        editableEnabled = false,
+        builderPackage = Constants.FABRIC8_KUBERNETES_API
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({"type", "valueFrom"})
+@EqualsAndHashCode
+public class HashLoginServiceApiUsers extends ApiUsers {
+    public static final String TYPE_HASH_LOGIN_SERVICE = "hashLoginService";
+
+    // Regex to match an entry in Jetty's HashLoginService's file format: username: password, rolename
+    public static final Pattern HASH_LOGIN_SERVICE_PATTERN = Pattern.compile("^[\\w-]+\\s*:\\s*\\w+\\s*,\\s*\\w+\\s*$");
```

Comment:
This is internal implementation detail -> it should be somewhere in the model class or in some utils. Same as the parsing methods.

Reply:
I think it could be private but why moving out from here? The way to parse entries is specific to each implementation of the ApiUsers, or? Its responsability is to know how to parse stuff.
On the other side, you could argue that this is part of the API module so visible to the users but then what does it mean, that we need a utils class to parse entries for each implementation of the ApiUsers?

Reply:
No, that is not the responsibility of the API. The API only provides the coordinates of the Secret that contains the configuration. The other implementation also might not have any users. They might be just an OAuth configuration for example. So the way this is abstracted is wrong as it is not abstract at all.

Reply:
> The API only provides the coordinates of the Secret that contains the configuration

ok but ...

> So the way this is abstracted is wrong as it is not abstract at all.

So which class is in charge of parsing the entries for a specific implementation? To me it should be the implementation itself for resposanbility segregation.

Reply:
Any possible parsing should happen in the `cluster-operator` module.

Reply:
But it would mean that for each API implementation of the ApiUsers you have a corresponding class in the operator module doing the parsing, or having just one Utils class with all parsing depending on the ApiUsers implementation. Not sure it's really good.

Reply:
Possibly. But you don't really know what you will be parsing - if anything. As I said, the next type might be a URL of an OAuth server for example without any _user list_ to parse. You cannot really prepare for it. 

I think the metrics implementation is a nice example of what you can do today in terms of abstraction. It is pretty similar, the API has just a Config Map link and all the parsing is handled in the `MetricsModel` class. When a new type is added, we will have to see how to deal with it. Maybe by adding stuff to the `MetricsModel` class, maybe creating some abstraction to it and having two implementations, maybe having it handled differently in the reconcilers. You do not really know it until you know the implementation details. If the new implementation is similar with first one and also provides some ConfigMap with the config it might be done differently than if it has few configuration flags directly in the CRD API etc.

Reply:
Yeah that's my fear. It's that you end having an abstraction on the api side, so for example class A implementing ApiUsers and class B implementing ApiUsers differently, and then you end with two other classes, let's say modelA and modelB which are in charge of doing the right parsing depending on the implementations of A and B.

Reply:
But why do you fear that? If you give users the options, the options might be different. There is nothing to fear on that, or? But I don't think it makes much sense to prepare for it now because you don't know how it will look like. You also don't know for sure if there ever will be more options. So preparing for possible future unknown options is just sunk cost. So my take is that we should make sure the API can be extended, but the internal code will be refactored when needed as we did for many other things in the same way ... authorization, authentication, listeners, storage, metrics, ...